includes/data-stores/class-wc-admin-reports-customers-data-store.php

```diff
@@ -0,0 +1,523 @@
+				'total'   => 0,
+				'pages'   => 0,
+				'page_no' => 0,
+			);
+
+			$selections       = $this->selected_columns( $query_args );
+			$sql_query_params = $this->get_sql_query_params( $query_args );
+
+			$db_records_count = (int) $wpdb->get_var(
+				"SELECT COUNT(*)
+					FROM
+						{$table_name}
+						{$sql_query_params['from_clause']}
+					WHERE
+						1=1
+						{$sql_query_params['where_time_clause']}
+						{$sql_query_params['where_clause']}
+				"
+			); // WPCS: cache ok, DB call ok, unprepared SQL ok.
+
+			$total_pages = (int) ceil( $db_records_count / $sql_query_params['per_page'] );
+			if ( $query_args['page'] < 1 || $query_args['page'] > $total_pages ) {
+				return $data;
+			}
+
+			$customer_data = $wpdb->get_results(
+				"SELECT
+						{$selections}
+					FROM
+						{$table_name}
+						{$sql_query_params['from_clause']}
+					WHERE
+						1=1
+						{$sql_query_params['where_time_clause']}
+						{$sql_query_params['where_clause']}
+					GROUP BY
+						customer_id
+					ORDER BY
+						{$sql_query_params['order_by_clause']}
+					{$sql_query_params['limit']}
+					",
+				ARRAY_A
+			); // WPCS: cache ok, DB call ok, unprepared SQL ok.
+
+			if ( null === $customer_data ) {
+				return $data;
+			}
+
+			$customer_data = array_map( array( $this, 'cast_numbers' ), $customer_data );
+			$data          = (object) array(
+				'data'    => $customer_data,
+				'total'   => $db_records_count,
+				'pages'   => $total_pages,
+				'page_no' => (int) $query_args['page'],
+			);
+
+			wp_cache_set( $cache_key, $data, $this->cache_group );
+		}
+
+		return $data;
+	}
+
+	/**
+	 * Add customer information to the lookup table when orders are created or modified.
+	 *
+	 * @param int $post_id Post ID.
+	 */
+	public static function sync_order( $post_id ) {
+		if ( 'shop_order' !== get_post_type( $post_id ) ) {
+			return;
+		}
+
+		$order = wc_get_order( $post_id );
+		if ( ! $order ) {
+			return;
+		}
+
+		$customer_id = $order->get_customer_id();
+		if ( 0 === $customer_id ) {
+			return;
+		}
+
+		self::update_registered_customer( $customer_id );
+	}
+
+	/**
+	 * Updates a guest (no user_id) customer data with new order data.
+	 *
+	 * @param WC_Order $order Order to update guest customer data with.
+	 * @return int|false The number of rows affected, or false on error.
+	 */
+	public function update_guest_customer_by_order( $order ) {
+		global $wpdb;
+
+		$email = $order->get_billing_email( 'edit' );
+
+		if ( empty( $email ) ) {
+			return true;
+		}
+
+		$existing_guest = $this->get_guest_by_email( $email );
+
+		if ( $existing_guest ) {
+			$order_timestamp  = date( 'Y-m-d H:i:s', $order->get_date_created( 'edit' )->getTimestamp() );
+			$new_orders_count = $existing_guest['orders_count'] + 1;
+			$new_total_spend  = $existing_guest['total_spend'] + (float) $order->get_total( 'edit' );
+
+			return $wpdb->update(
+				$wpdb->prefix . self::TABLE_NAME,
+				array(
+					'orders_count'     => $new_orders_count,
```

Comment:
Since we need to add the `customer_id` to the order stats table, I'm wondering if we should just perform a JOIN here on the order stats to get the current information.

We can then grab `orders_count`, `total_spend`, `avg_order_value`, and `date_last_order` without needing to update them here regularly.

I think this would result in less syncing issues and less overhead here.  Thoughts?

Reply:
_Restating to make sure I've understood:_

1. At the time of guest order creation (and therefore update to the `wc_order_stats` table) we'd create a customer in the `wc_customer_lookup` table if none exists for the billing email
2. When hitting the `reports/customers` endpoint, the values for `orders_count`, `total_spend`, `avg_order_value`, and `date_last_order` would come from a `JOIN` on the `wc_order_stats` table, and all syncing code for `wc_customer_lookup` can be removed

Reply:
The solution you outlined above @jeffstieler seems reasonable to me, and will ensure the most current data about a customer for the given query to the REST API. Also seems like it should be performant.

Reply:
Looks good to me, @jeffstieler 👍